pkg/schedule/schedulers/balance_range.go

```diff
@@ -167,6 +227,332 @@ func newBalanceRangeScheduler(opController *operator.Controller, conf *balanceRa
+	}
+	candidates := filter.NewCandidates(s.R, dstStores).FilterTarget(conf, nil, s.filterCounter, filters...)
+	for i := range candidates.Stores {
+		plan.target = candidates.Stores[len(candidates.Stores)-i-1]
+		plan.targetScore = plan.score(plan.target.GetID())
+		if plan.targetScore > plan.averageScore {
+			break
+		}
+		regionID := plan.region.GetID()
+		sourceID := plan.source.GetID()
+		targetID := plan.target.GetID()
+		if !plan.shouldBalance(s.GetName()) {
+			continue
+		}
+		log.Debug("candidate store", zap.Uint64("region-id", regionID), zap.Uint64("source-store", sourceID), zap.Uint64("target-store", targetID))
+
+		oldPeer := plan.region.GetStorePeer(sourceID)
+		exist := false
+		if plan.job.Role == leader {
+			peers := plan.region.GetPeers()
+			for _, peer := range peers {
+				if peer.GetStoreId() == targetID {
+					exist = true
+					break
+				}
+			}
+		}
+		var op *operator.Operator
+		var err error
+		if exist {
+			op, err = operator.CreateTransferLeaderOperator(s.GetName(), plan, plan.region, plan.targetStoreID(), []uint64{}, operator.OpRange)
+		} else {
+			newPeer := &metapb.Peer{StoreId: plan.target.GetID(), Role: oldPeer.Role}
+			op, err = operator.CreateMovePeerOperator(s.GetName(), plan, plan.region, operator.OpRange, oldPeer.GetStoreId(), newPeer)
+		}
+
+		if err != nil {
+			balanceRangeCreateOpFailCounter.Inc()
+			return nil
+		}
+		sourceLabel := strconv.FormatUint(sourceID, 10)
+		targetLabel := strconv.FormatUint(targetID, 10)
+		op.FinishedCounters = append(op.FinishedCounters,
+			balanceDirectionCounter.WithLabelValues(s.GetName(), sourceLabel, targetLabel),
+		)
+		op.SetAdditionalInfo("sourceScore", strconv.FormatInt(plan.sourceScore, 10))
+		op.SetAdditionalInfo("targetScore", strconv.FormatInt(plan.targetScore, 10))
+		op.SetAdditionalInfo("tolerate", strconv.FormatInt(plan.tolerate, 10))
+		return op
+	}
+	balanceRangeNoReplacementCounter.Inc()
+	return nil
+}
+
+// balanceRangeSchedulerPlan is used to record the plan of balance key range scheduler.
+type balanceRangeSchedulerPlan struct {
+	sche.SchedulerCluster
+	// stores is sorted by score desc
+	stores []*core.StoreInfo
+	// scoreMap records the storeID -> score
+	scoreMap     map[uint64]int64
+	source       *core.StoreInfo
+	sourceScore  int64
+	target       *core.StoreInfo
+	targetScore  int64
+	region       *core.RegionInfo
+	fit          *placement.RegionFit
+	averageScore int64
+	job          *balanceRangeSchedulerJob
+	opInfluence  operator.OpInfluence
+	tolerate     int64
+}
+
+type storeInfo struct {
+	store *core.StoreInfo
+	score int64
+}
+
+func (s *balanceRangeScheduler) prepare(cluster sche.SchedulerCluster, opInfluence operator.OpInfluence, job *balanceRangeSchedulerJob) (*balanceRangeSchedulerPlan, error) {
+	krs := core.NewKeyRanges(job.Ranges)
+	scanRegions, err := cluster.BatchScanRegions(krs)
```

Comment:
Do we need to separate krs into multiple batches if it is a big range?

Reply:
It doesn't need to separate the ranges by many times because the duration of the lock is the same

Reply:
If we acquire the lock multiple times and for every time we hold the lock for a short time, the other places can also get the lock. But If we hold the lock for a long time, it may cause other places starvation.

Reply:
ok, I will split the ranges.